lib/rucio/common/plugins.py

```diff
@@ -74,12 +76,44 @@ class PolicyPackageAlgorithms:
     """
     _ALGORITHMS: dict[str, dict[str, 'Callable[..., Any]']] = {}
     _loaded_policy_modules = False
+    _default_algorithms: dict[str, 'Callable[..., Any]'] = {}
 
     def __init__(self) -> None:
         if not self._loaded_policy_modules:
             self._register_all_policy_package_algorithms()
             self._loaded_policy_modules = True
 
+    @classmethod
+    def _get_default_algorithm(cls: type[PolicyPackageAlgorithmsT], algorithm_type: str, vo: str = "") -> Optional['Callable[..., Any]']:
+        """
+        Gets the default algorithm of this type, if present in the policy package.
+        The default algorithm is the function named algorithm_type within the module named algorithm_type.
+        Returns None if no default algorithm present.
+        """
+        # check if default algorithm for this VO is already cached
+        type_for_vo = algorithm_type + "-" + vo
+        if type_for_vo in cls._default_algorithms:
+            return cls._default_algorithms[type_for_vo]
+
+        default_algorithm = None
+        try:
+            if vo == 'def':
+                vo = ''
+            package = cls._get_policy_package_name(vo)
+        except (NoOptionError, NoSectionError):
+            return default_algorithm
+
+        try:
+            module_name = package + "." + algorithm_type
+            module = importlib.import_module(module_name)
+
+            if hasattr(module, algorithm_type):
+                default_algorithm = getattr(module, algorithm_type)
+                cls._default_algorithms[type_for_vo] = default_algorithm
+        except ImportError:
+            LOGGER.warning('Policy package %s could not be loaded' % package)
```

Comment:
From https://docs.python.org/3/library/exceptions.html


exception ImportError

Raised when the [import](https://docs.python.org/3/reference/simple_stmts.html#import) statement has troubles trying to load a module. Also raised when the “from list” in from ... import has a name that cannot be found.


So I think we should log `module` instead of `package`, and say that the module could not be imported

Reply:
Thinking about it some more, I'm not actually sure if we should log this at all, except maybe at a very low level such as info or debug. This exception will happen if the module containing one of the default algorithms can't be loaded, but that's normal and expected if the package doesn't implement that algorithm. If the policy package itself can't be loaded then that should be picked up and logged elsewhere.